parallelize getting mount info

bcoca opened the pull request:
try to mitigate slow gather_facts by parallelizing the #1 source of delays

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
setup

bcoca committed:
parallelize getting mount info

nitzmahone commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -483,7 +472,22 @@ def get_mount_facts(self):
+        uuids = self._lsblk_uuid()
+        mtab_entries = self._mtab_entries()
+
+        pool = ThreadPool(processes=len(mtab_entries))
```
should probably have an upper bound of < 10 just for sanity

bcoca committed:
fixed timeout and made 8 max thread count

  - minor cleanup
  - avoid empty mount entries
  - set timeout on get

abadger commented:
Keeping track of when we first invoked queue.get() would allow us to reduce the timeout period.  The problem is that all of the calculations aren't truly done in parallel.  We're limited by number of threads that we start and number of processor cores on the box to handle those threads. So we can't know whether a particular query started the first time we invoked queue.get it later.  Perhaps that doesn't matter or perhaps we should do it own pool management...

abadger commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -483,7 +471,28 @@ def get_mount_facts(self):
+        # start threads to query each mount
+        mounts = []
+        async_results = []
+        pool = ThreadPool(processes=min(len(mtab_entries), 8))
```
If you're trying to make sure we don't overtax the cpu, use this: ```min(len(mtab_entries), multiprocessing.cpu_count())```

bcoca replied:
done

abadger commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -483,7 +471,28 @@ def get_mount_facts(self):
-            mounts.append(mount_info)
+            return mount_info
+
+    @timeout.timeout()
```
No longer need timeout, I think because we'll use timeouts from multiprocessing.

bcoca replied:
both, keep a global one for rest of function

bcoca replied:
but on the other hand, its same length of time .. so it might not make sense anymore, since we do it by mount vs for all mounts

abadger commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -443,25 +443,13 @@ def _mtab_entries(self):
-
-            if fstype == 'none':
-                continue
+            if all([not device.startswith('/'), ':/' not in device, not fstype.startswith('fuse')]) or fstype == 'none':
```
It's more readable to write this out into multiple if-then statements.

bcoca replied:
moved and reverted

abadger commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -443,25 +443,13 @@ def _mtab_entries(self):
-        mount_facts = {}
-
-        mount_facts['mounts'] = []
+    def get_mount_info(self, fields, bind_mounts, uuids):
```
This can become a function since it no longer makes use of self.

bcoca replied:
has still 2 uses of self

bcoca committed:
add nuance

  - enforce timeout per mount/thread
  - make note on failure per mount
  - make note on timeout per mount
  - ensure proper pool control
  - minor fixes
  - less vars, simpler code
  - move filter 'pre threading'

bcoca committed:
remove timeout for all mounts, now per mount

  also use cpu count from multiprocessing lib

bcoca committed:
peeepeee

abadger commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -443,51 +445,82 @@ def _mtab_entries(self):
+                # only add if not already there, we might have a plain /etc/mtab
+                if not self.MTAB_BIND_MOUNT_RE.match(options):
+                    options += ",bind"
+            mount_info['options'] = options
```
This piece on mount should be moved to the outer level as it isn't in danger of timing out and will then always be provided to the caller (whereas if it's in here, it will be omitted if something else fails).

bcoca replied:
done

abadger commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -443,51 +445,82 @@ def _mtab_entries(self):
-    @timeout.timeout()
+    def get_mount_info(self, mount_info, bind_mounts, uuids):
+
+            mount_info.update(get_mount_size(mount_info['mount']))
```
Don't modify things inside of a function unless there's a huge performance hit.  Just return the values.

Also, don't pass in a meta object like mount_info...  Instead be explicit that you need the mount and the device by passing just those two arguments in.

Looking at this further, though, it probably makes more sense to call each of those separately as they're disjoint information and could timeout separately.

bcoca replied:
refactored as per comments, keeping both operations in same thread for now, since if the mount is going to be problematic, both of those would trigger the same problem. 

Also i would abstract the threading a bit vs having to code it for every possible function that can cause stalls/timeouts. First i want to see if this actually does improve those situations though.

bcoca committed:
moved 'bind' options out of thread as per comments

abadger commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -443,51 +445,81 @@ def _mtab_entries(self):
+            results[mount] = {'info': mount_info,
+                              'extra': pool.apply_async(self.get_mount_info, (mount_info, uuids)),
+                              'start': time.time() + maxtime}
+
```
You don't need start with this code as it will be virtually the same for all the things launched by apply_async in your loop.  You can just use a single scalar local variable to hold the value.  (It also is the end time or max time (and maxtime is really max_timeout), not start).

Adding that together with the note that we should probably process statvfs information separate from the uuid info:
``` python
results[mount] = {'info': mount_info,
                            'statvfs': pool.apply_async(get_mount_size, (mount_info['mount'],)),
                            'uuid': uuids.get(mount_info['device']) or pool.apply_async(self._udevadm_uuid, (mount_info['device'],)),}
max_time = time.time() + max_timeout
```

bcoca replied:
not as single scalar since we are tracking timeout PER child , otherwise we could just leave the general timeout at top of the method

abadger replied:
You are not tracking timeout per child with this code.  The  time you're recording here is when you ask the library to queue the work.... It's not when the work starts.  Therefore, you might as well save a single general timeout at the top of the method.

bcoca replied:
well, yes, queueing per child to be exact, in some of my most extreme tests this was up to 2secs diff, so i opted to keep it per child

bcoca replied:
I was looking at time.clock() at the begining of the threaded function as that would give us a more accurate 'execution time'  but we cannot compare it before the `.get` which is the thing we skip on timeout.

abadger replied:
What was your extreme test?

bcoca replied:
hosts with +300 mount points

abadger replied:
I think you should rerun your test... The cost to invoke apply_async() on 300 elements is .01 seconds here. Adding in the regex, it's still only .05 seconds.  .05 seconds on a 10 second timeout is virtually nothing.

bcoca replied:
Close to what i see when its a single target, but i'm testing on 50-70 vms on same host (2cpu) with variable mount points (shell below creates them) in parallel, which is not a fair test. 

I was trying to find a point in which i could check the difference in performance between threaded/non threaded and I ended up with this, which also showed me the queuing diff can get big on busy system. The check themselves need more variety, seems like bind mounts end up caching the info making it a non optimal test, still trying to figure out something i can whip up locally.

```bash
#!/bin/bash

ADD=$((${RANDOM}%100))
LIM=$((250+${ADD}))

for i in $(seq 1 $LIM)
do
        mkdir /mnt/test_$i
        mount -o bind /var /mnt/test_$i
done
```

bcoca committed:
refactored thread function

bcoca committed:
changed start to 'timelimit', more accurate

 better reflects current use, as per feedback

bcoca committed:
warn on error, more info on failure to get info

  also fix func sig

bcoca committed:
more sensible diff limit

abadger commented on lib/ansible/module_utils/facts/hardware/linux.py:
```diff
@@ -443,51 +446,85 @@ def _mtab_entries(self):
+                    results[mount]['info']['note'] = 'Timed out while attempting to get extra information.'
+                    mounts.append(results[mount]['info'])
+                    del results[mount]
+                    break
```
@nitzmahone and I talked about this last night and the timeout is an expectation of how long fact gathering as a whole will take, not how long any specific piece of fact gathering will take.  We thought that probably the best place to enforce that sort of fact gathering would be in the action plugin being worked on in a separate PR.  However, for this PR, it would make more sense to at least try to replicate the status quo as much as possible, ie: take a time at the start of this function.  Then every time we check the time to see if it's time to exit due to timeout, we compare time.time() > function_start_time + timeout to see if it's time to exit.

bcoca replied:
from the docs on setup module
```
- gather_timeout
        Set the default timeout in seconds for individual fact gathering
```

bcoca replied:
as to what 'individual facts' are we never define, so here i was using the timeout per fact, but running them in parallel as much as possible, the total might take a bit more but it should be close enough. We read all ohter facts serially, but we can improve that with this kind of code going forward.

bcoca committed:
removed redundant terminate and harmful join
